src/planet-a/SubdivSkeleton.h

```diff
@@ -0,0 +1,392 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/id_registry.h>
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+#include <optional>
+#include <limits>
+#include <unordered_map>
+#include <vector>
+
+namespace planeta
+{
+
+template<typename T>
+using ArrayView_t = Corrade::Containers::ArrayView<T>;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A multitree directed acyclic graph of reusable IDs where new IDs can
+ *        be created from two other parent IDs.
+ */
+template<typename ID_T>
+class SubdivIdTree : private osp::IdRegistry<ID_T>
+{
+
+    using base_t = osp::IdRegistry<ID_T>;
+    using id_int_t = std::underlying_type_t<ID_T>;
+
+    static_assert(std::is_integral<id_int_t>::value && sizeof(ID_T) <= 4,
+                  "ID_T must be an integral type, 4 bytes or less in size");
+
+    using combination_t = uint64_t;
+
+public:
+
+    using base_t::capacity;
+    using base_t::size;
+
+    ID_T create_root()
+    {
+        ID_T const id = base_t::create();
+        m_idChildCount.resize(capacity());
+        m_idChildCount[size_t(id)] = 0;
+        return id;
+    };
+
+    std::pair<ID_T, bool> create_or_get(ID_T a, ID_T b)
+    {
+        combination_t const combination = hash_id_combination(a, b);
+
+        // Try emplacing a blank element under this combination of IDs, or get
+        // existing element
+        auto const& [it, success] = m_parentsToId.try_emplace(combination, 0);
+
+        if (success)
+        {
+            // The space was free, and a new element was succesfully emplaced
+
+            // Create a new ID for real, replacing the blank one from before
+            it->second = id_int_t(create_root());
+
+            // Keep track of the new ID's parents
+            m_idToParents.resize(capacity());
+            m_idToParents[it->second] = combination;
+
+            // Increase child count of the two parents
+            m_idChildCount[size_t(a)] ++;
+            m_idChildCount[size_t(b)] ++;
```

Comment:
wouldn't you want pre-incremener here?

Reply:
Is this for readability? I feel that it makes more sense that the increment happens after the [] operator.